### ISSUE TITLE
Plugin chat prompts are no longer being cancelled from being visible in chat

### ISSUE THREAD

**Issue by Sikatsu** [PaperSpigot/Paper#8911]
### Expected behavior

When a plugin asks me to type a value within the chat, the message should be cancelled from being visible to everyone.

### Observed/Actual behavior

The messages are no longer being cancelled and are visible to every player on the server.

### Steps/models to reproduce

You can try this with many different plugins, as they are all broken in this regard:

- Auction house
- DisplayShops
- HyperDrive
- HeadDatabase
- PremiumVanish/SuperVanish
- etc

The issue is there after updating from paper build 399 to 420.
I have tested this on my survival, creative AND nearly empty test server with just HDB and the two different paper builds.

### Plugin and Datapack List

Plugins (28): AntiPopup, ArmorStandEditor, ArtMap, CMI, CMILib, ChatControlRed, CommandPanels, CoreProtect, ExecuteEverywhere, HeadDatabase, LuckPerms, Multiverse-Core, MyCommand, PlaceholderAPI, PlayerParticles, PlotSquared, PremiumVanish, ProtocolLib, ServerUtils, ShowItem, SupremeTags, TAB-Bridge, VCLink2-Bukkit, Vault, VillagerCraftCore, WorldEdit, WorldGuard and eGlow

### Paper version

[11:15:16 INFO]: This server is running Paper version git-Paper-420 (MC: 1.19.3) (Implementing API version 1.19.3-R0.1-SNAPSHOT) (Git: 81d7ff6)
You are running the latest version

### Other

![image](https://user-images.githubusercontent.com/5664135/221153329-75d5b43b-ca5b-4080-87a2-d5ed19c5c265.png)

**Comment by electronicboy** [comment]
Hard to say around premium/closed source plugins, given that the only real change has been in the plugin loader, this would generally scream more towards a plugin issue, primarily around their dependency on event order;

**Comment by Sikatsu** [comment]
I see. Wouldn't it be suspicious this happens to all plugins though? None of the ones I mentioned are open source?

For now i'll stay on the old version.

**Comment by Sikatsu** [comment]
It indeed runs behind a waterfall proxy. Though nothing was changed or updated on that regard. All that I have actually changed is the paper version on the backend servers.

I'll take a look at the website you provided.

**Comment by Sikatsu** [comment]
Would installing HuskChat also solve my issue?

On a clean bungee installation with 2 backend servers, running waterfall and paper with vault, luckperms and hyperdrive I am able to reproduce this issue unless I rollback my paper version to an older build.

**Comment by electronicboy** [comment]
Would need to isolate what build broke it, as that makes zero sense to me

**Comment by Sikatsu** [comment]
Not sure which build exactly, but between 399 and 420. I could try to pinpoint it exactly, but unnfortunately that has to be for monday as i'm out for a few days.

**Comment by Sikatsu** [comment]
@electronicboy I can confirm that build 404 still works, it breaks on build 405 for ALL plugins that work with a chat prompt.

With this commit: https://github.com/PaperMC/Paper/issues/8108

**Comment by Omarlatif** [comment]
I can confirm that on my server it also breaks on build 405.

**Comment by MiniDigger** [comment]
does somebody have code to repro? I am having a hard time trying to understand what you mean by prompts, is it the conversation api that is broken? or just plugins cancelling one of the chat events?

**Comment by electronicboy** [comment]
They're using chhat events, not the conversation API; this generally looks
like a case of plugins conflicting over priorities and load order

**Comment by NyhmsQuest** [comment]
This is happening on my server too with all my plugins using chat prompts.
Something changed or broke with that Paper update

**Comment by Owen1212055** [comment]
Please try to provide as simple reproduction steps if possible, as generally, this issue can be a wide variety of things due to how much truly changed with Paper plugins.

This would allow us to do much better debugging and figure out the root cause of the issue here.

**Comment by kangarko** [comment]
Similar problem also happens with my plugin ChatControl Red, 

https://github.com/kangarko/ChatControl-Red/issues/2306

Paper needs to be more clear on the changes and provide developers instructions of what is wrong with their plugins, and what they need to change specifically.

Let me know your BuiltByBit ID and I can grant you license to ChatControl Red if you need to debug it. Or let me know where to find the culprit and I am happy to collaborate on this.

**Comment by Sikatsu** [comment]
The reproduction steps are quite easy once you have a plugin that makes use of chat prompts. Unfortunately can't provide a simple plugin as that's not my expertise. I also make use of ChatControl Red, so testing with Kangarko's plugin as mentioned above might be an option?

**Comment by Yertled** [comment]
ChatSentry and ExecutableItems are 2 plugins that we notice this issue with the hardest.
ExecutableItems have a free version:
https://www.spigotmc.org/resources/custom-items-free-executable-items.77578/

Download the plugin and run it on the server. Use the broken Paper build (405), or in our case, latest PurpurMC build.
Then create an item with Executable Items.
Write in chat:
/ei create TEST
Then click on the "Edit Name" or "Edit Lore"
It would allow you to type the name of the item, or something random for the lore in chat. Instead of cancelling what you type to chat, it publically broadcasts what I type, even tho it should be cancelled.

This has worked perfectly for months, until we updated our Purpur (which is Paper).

**Comment by Brennian** [comment]
From what I have heard, this doesn't work anymore.
```
    @EventHandler(priority = EventPriority.LOW)
    public void onChat(AsyncPlayerChatEvent e) {
        if (e.isCancelled()) {
            return;
        }
        if (enabledSearchNames.contains(e.getPlayer().getName())) {
            openSearchInventory(e.getMessage(), 1, e.getPlayer());
            enabledSearchNames.remove(e.getPlayer().getName());
            e.setCancelled(true);
        }
    }
```

**Comment by Owen1212055** [comment]
> Similar problem also happens with my plugin ChatControl Red, 
> 
> 
> 
> https://github.com/kangarko/ChatControl-Red/issues/2306
> 
> 
> 
> Paper needs to be more clear on the changes and provide developers instructions of what is wrong with their plugins, and what they need to change specifically.
> 
> 
> 
> Let me know your BuiltByBit ID and I can grant you license to ChatControl Red if you need to debug it. Or let me know where to find the culprit and I am happy to collaborate on this.

Our guess is that this might have something to do with load order. Because recent changes have not touched the event system at all. But, we are not sure.

**Comment by DaJokni** [comment]
> For example in headdatabase there is a button in the GUI ''Search heads..'', when I click on it, a title comes up to enter a value within the chat. When I do that however, the value is visible to everyone on the server, rather then being cancelled.
> 
> Any plugin that makes use of chat prompts (enter a value in chat, like prices etc) is broken as of build 405.
> 
> Don't really have any code myself to reproduce, as I am not a developer unfortunately.

do you mean the search function in hdb? just tested that on paper build 445 and it appears to be working correctly

**Comment by Sikatsu** [comment]
No, not the regular search function. That works fine. The one in the GUI which after you click on it, asks you to typ your input in chat. Could be that you have that option disabled. It was just an example of a use case though.

**Comment by Sikatsu** [comment]
It works, but it doesn't cancel the message for the whole server to see. 
I can reproduce my issue on a clean test server running paper build 445 as well..

**Comment by DaJokni** [comment]
> It works, but it doesn't cancel the message for the whole server to see. I can reproduce my issue on a clean test server running paper build 445 as well..

it works, and the message isnt sent to the whole server. nothing seems out of the ordinary

**Comment by kangarko** [comment]
Oh by the way, ChatControl, one of the affected plugins, is sending chat messages as system messages (we use Spigot API for compatibility purposes). I assume this might be a contributing factor?

**Comment by kangarko** [comment]
@emilyy-dev can it be that we're using EventExecutor instead of Listener for the chat event specifically? https://github.com/kangarko/Foundation/blob/master/src/main/java/org/mineacademy/fo/event/SimpleListener.java

**Comment by Sikatsu** [comment]
What's the progress on this one? We can still not update beyond build 404, even if we want to update to 1.19.4?

**Comment by Machine-Maker** [comment]
We need an example of this issue using something that we can actually test and interact with and see what the plugin is doing. So far, all reproduction steps have involved a closed source plugin which means we can't easily see what interactions might be causing this. There are different ways plugins can implement such a system of responding to prompts and we don't know which one might be the issue.

**Comment by kangarko** [comment]
[ChatControl-Red-10.17.4.zip](https://github.com/PaperMC/Paper/files/11021666/ChatControl-Red-10.17.4.zip)

@Machine-Maker here is my premium plugin for debug purposes. I have also invited you to [ChatControl repository](https://github.com/kangarko/chatcontrol-red-src). We compile the plugin on top of [this library](https://github.com/kangarko/foundation).

**Comment by Machine-Maker** [comment]
Ok so without any other plugins, the prompt functionality for the `/mail` command's forward feature does work. I was able to just type in my own player name and the forwarding worked and the message wasn't sent in chat. So that points to it being a conflict with another plugin (probably load order related).

**Comment by Sikatsu** [comment]
> @Sikatsu can you try opening ChatControl.jar with WinRar or similar and editing its plugin.yml > try removing or adding your conflicting plugin from/to the soft-depend key list to see if any changes occur

I could try that, however it happens with ALL plugins, even ones that are not in the soft-depended list..

**Comment by electronicboy** [comment]
Nothing has changed at all with how events work, this is once again likely to be an order of execution issue, a quick plugin to print out the event handlers order might be an idea, but, I don't recall if enough useful info is retained in the executors to provide anything useful

**Comment by sandwally** [comment]
Those interactions works fine with chatcontrol, it's when you change the settings.yml for chatcontrol to enable channels as chat formatting it starts to pass through both the text to chat and it executes the chat prompt command.

**Comment by kangarko** [comment]
@Machine-Maker thank you for taking the time to test users' code.

I have made that quick plugin to print out registered listeners: 
[PrintEvents.zip](https://github.com/PaperMC/Paper/files/11060992/PrintEvents.zip)

With CMI, ChatControl and ItemsAdder this is the order in which they are loaded:
![2023-03-24-11 24-208](https://user-images.githubusercontent.com/1918426/227496175-e93ae534-3114-4ffc-87bd-bec95dfde326.png)

I unfortunately have no idea from the data provided how to fix this and would appreciate any direction. @Sikatsu I assume you tried changing our Listener_Priority from LOWEST to HIGHEST in ChatControl, right?

**Comment by Sikatsu** [comment]
It's always been on HIGHEST for me.

**Comment by kangarko** [comment]
If if its on HIGHEST then by theory we should catch the message as last, thus allowing CMI to properly cancel the event since we ignore canceled events.

@electronicboy can you check if truly nothing has changed for legacy event registration which lets us specify ignoreCancelled flag as well as priority? This is what I am referring to that we are using -- if the plugin with a lower priority cancels the event then we should just not get the method fired under the latest Paper version, or I am mistaken? It worked like this previously.

![a](https://i.imgur.com/GPSrxhF.png)

**Comment by Sikatsu** [comment]
I'm not sure if CMI makes use of chat inputs personally, but it's definitely on highest and causing issues with plugins I mentioned in my original message.

Thanks for looking into it, waiting on this one before I can update bug-free to 1.19.4

**Comment by Sikatsu** [comment]
I do not think it does, however I disabled CMI's chat entirely. Not sure if that matters at all?

**Comment by kangarko** [comment]
I don't know how to proceed with this. I provided the requested information to Paper dev. I gave the source code to @Machine-Maker. I know we are not the only plugin causing issues like this I truly believe something must have changed in Paper. I am open and happy to collaborate on fixing this.

**Comment by Sikatsu** [comment]
I'm afraid this issue is not being seen though and it's kind of holding me back to update to 1.19.4 as well, just because this bug will go live if I do so. Is there not a way to check the paper commits or try a different method for this? As without ChatControl, it does indeed seem to work..

**Comment by electronicboy** [comment]
I've already said what the issue probably is; nothing in paper has changed
with the event system, cancelled is still literally just aa simple boolean;
this is going to be a plugin issue down to the order that event handlers
are executed, I'd imagine thhat if you use the flag to use the older
dependency order calculations that it might just magically work again (ofc,
not a solution)

**Comment by Sikatsu** [comment]
I tried using the flag, but it seemed to break a couple of plugins for me. For example chat tags/badges could not be parsed and just showed the placeholder in chat.

Hope kangarko has an idea

**Comment by kangarko** [comment]
I'm going to try to create a sample plugin combination to debug this, please allow extra 2-5 days as I have other works right now.

**Comment by Sikatsu** [comment]
It's definitely related to ChatControl, we had someone create a test plugin as well which works unless ChatControl is installed. Hope for a fast fix so we can update to 1.19.4..

**Comment by Owen1212055** [comment]
This most likely has been resolved in https://github.com/PaperMC/Paper/pull/9099.

Please feel free to leave a comment if you are still experiencing this issue.
